x/evm/keeper/state_transition.go

```diff
@@ -515,81 +592,276 @@ func (k *Keeper) ApplyMessageWithConfig(
-	var replyLog StateDBGetLogsReply
-	err = sgxRPCClient.StateDBGetLogs(StateDBGetLogsArgs{}, &replyLog)
+	respLogs, err := sgxGrpcClient.StateDBGetLogs(ctx, &sgxtypes.StateDBGetLogsRequest{HandlerId: handlerId})
 	if err != nil {
+		// panic cosmos if sgx isn't available.
+		if k.IsSgxDownError(err) {
+			panic("sgx rpc server is down")
+		}
 		return nil, err
 	}
 
+	ethlogs := make([]*ethtypes.Log, 0)
+	for _, log := range respLogs.Log {
+		ethLog := &ethtypes.Log{
+			Address: common.BytesToAddress(log.Address),
+			// list of topics provided by the contract.
+			// supplied by the contract, usually ABI-encoded
+			Data: log.Data,
+
+			// Derived fields. These fields are filled in by the node
+			// but not secured by consensus.
+			// block in which the transaction was included
+			BlockNumber: log.BlockNumber,
+			// hash of the transaction
+			TxHash: common.BytesToHash(log.TxHash),
+			// index of the transaction in the block
+			TxIndex: uint(log.TxIndex),
+			// hash of the block in which the transaction was included
+			BlockHash: common.BytesToHash(log.BlockHash),
+			// index of the log in the block
+			Index: uint(log.Index),
+
+			// The Removed field is true if this log was reverted due to a chain reorganisation.
+			// You must pay attention to this field if you receive logs through a filter query.
+			Removed: log.Removed,
+		}
+
+		topics := make([]common.Hash, 0)
+		for _, topic := range log.Topics {
+			topics = append(topics, common.BytesToHash(topic))
+		}
+
+		ethLog.Topics = topics
+		ethlogs = append(ethlogs, ethLog)
+	}
+
 	return &types.MsgEthereumTxResponse{
 		GasUsed:   gasUsed,
 		VmError:   vmError,
 		Ret:       ret,
-		Logs:      types.NewLogsFromEth(replyLog.Logs),
+		Logs:      types.NewLogsFromEth(ethlogs),
 		Hash:      cfg.TxConfig.TxHash.Hex(),
 		BlockHash: ctx.HeaderHash(),
 	}, nil
 }
 
+func (k *Keeper) SafeBigInt2Uint64Convert(val *big.Int) uint64 {
+	if val == nil {
+		return 0
+	}
+	return val.Uint64()
+}
+
+func (k *Keeper) ChainConfigBitInt2Uint64(val *big.Int) uint64 {
+	if val == nil {
+		return 99999
+	}
+	return val.Uint64()
+}
+
+func (k *Keeper) SafeAddress2Bytes(addr *common.Address) []byte {
+	if addr == nil {
+		return nil
+	}
+	return addr.Bytes()
+}
+
+func (k *Keeper) prepareSgxChainConfig(cfg *EVMConfig) sgxtypes.ChainConfig {
+	chainConfig := cfg.ChainConfig
+	sgxChainConfig := sgxtypes.ChainConfig{
+		// *big.Int
+		ChainID: k.SafeBigInt2Uint64Convert(chainConfig.ChainID),
+		// *big.Int
+		HomesteadBlock: k.ChainConfigBitInt2Uint64(chainConfig.HomesteadBlock),
+		// *big.Int
+		DAOForkBlock:   k.ChainConfigBitInt2Uint64(chainConfig.DAOForkBlock),
+		DAOForkSupport: chainConfig.DAOForkSupport,
+		// EIP150 implements the Gas price changes (https://github.com/ethereum/EIPs/issues/150)
+		// *big.Int
+		EIP_150Block: k.ChainConfigBitInt2Uint64(chainConfig.EIP150Block),
+		// *big.Int
+		EIP155Block: k.ChainConfigBitInt2Uint64(chainConfig.EIP155Block),
```

Comment:
What's your opinion @kenta-mori3322 on keeping this code (that I suggested) versus using `bytes` and `json.Marshal` (your initial version)?

Maybe I was wrong, and your initial idea is more readable. What do you think?

Reply:
@amaury1093 I think this brought a lot complexity actually. We could find json.Unmarshal from existing code like EstimateGas in qrpc_query.go. I am spending much time on type conversion and it seems like it still has bug.
![Uploading data.png…]()

Reply:
OK. Let's revert to json.Marshal in this case, if this brings complexity. We can use json.Marshal for:
- core.Message
- rules
- overrides
- EVMConfig and ChainConfig

do you agree?

Reply:
@amaury1093 Agreed.